### PR TITLE
Various optimizations for the DataFlowAnalyzer.

### PR DESCRIPTION
**Pull request by ekpyron**
Part of #10405

I'm not necessarily saying that any of this is a particularly good idea :-)...
But with this I get a speedup of almost 15% on a RelWithDebInfo build:

develop:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m47,327s
user    0m46,901s
sys     0m0,294s
```

this branch:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m40,649s
user    0m40,286s
sys     0m0,259s
```

But I'm not sure how much further one can get with micro-optimizations like this.

### CODE REVIEW

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [inline libyul/optimiser/DataFlowAnalyzer.cpp:301]: This here is the single most serious change in the PR.
``clearValues`` is among the most expensive operations, especially since it has to clear back-references.
However, while clearing back-references is absolutely necessary e.g. for assignments, at the end of scopes we don't really have to care that much (joining knowledge on branches is handled differently). The only downside of not clearing here is that the ``DataFlowAnalyzer`` might report matches that are no longer in scope. However, all users/derived classes of ``DataFlowAnalyzer`` already had an additional assertion ``inScope(var)``, whenever a reported match was about to be used. These assertions are just turned into conditions in this PR, allowing us to skip the clearing here.

It's still questionable, of course, but this is the single most effective change here.

**Comment by ekpyron** [comment]
The most expensive remaining part when I profile this branch is ``DataFlowAnalyzer::handleAssignment``, so if there is much further to be done that's where to look.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: What if we don't clear back references here, but instead just remove the variables? The upside of clearing at the end of a scope is also that the data structures remain smaller.

File: libsolutil/InvertibleMap.h
Comment (chriseth) [inline libsolutil/InvertibleMap.h]: Does the remaining code actually warrant its own class? The `fetch` function could be helpful in the global scope, maybe also providing a default value if the key is not present. `eraseValue` could also be helpful in the global scope, also if not that much.

File: libyul/Dialect.h
Comment (chriseth) [inline libyul/Dialect.h]: Can you extract this into its own PR?

File: libyul/optimiser/CommonSubexpressionEliminator.cpp
Comment (chriseth) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp]:
```suggestion
			if (SyntacticallyEqual{}(_e, *value.value) && inScope(variable))
```

File: libsolutil/InvertibleMap.h
Comment (ekpyron) [reply libsolutil/InvertibleMap.h]: Yeah, it's mainly still here for easier rapid-changing - if we settle on any version of this, the class can be removed.
The ``fetch`` thing I'm actually not yet sure about - I have to remeasure performance on it and it only really works well on small cheaply-copiable value types... it would actually have to return ``std::optional<V const&>``, but unfortunately there's no such thing... we could go for a pointer, though...

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: Yes, I wondered as well, but haven't tried yet.

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: I'm also in general not yet 1000% sure that it's true that not clearing and instead checking on the user side doesn't have any unwanted side-effects... Tests pass, though, at least...

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: Actually, we could replace these with ``std::erase_if``, that's even specified to do the same for maps... https://en.cppreference.com/w/cpp/container/map/erase_if

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Ah, or damn, no, we can't - those are C++20 :-).

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: We should probably come up with a nice way to add functions like that without running into ambiguity problems later... i.e. something like a ``cxx20`` namespace which we never import, but always only use fully qualified... and which can then later just be dropped... or something like that... and there's the question if we do it ourselves whereever we want or if we want to pull in some header-only libraries that already does it extensively... Not sure how good e.g. https://abseil.io/ or similar projects are...

Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: yep sounds good. Either boost or our own implementation.

**Comment by axic** [comment]
What is the status of merging this?

**Comment by ekpyron** [comment]
> What is the status of merging this?

We'll probably merge some version at some point, but some comments etc. still need to be addressed. My plan was to take care of this one tomorrow.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m40,954s
user    0m40,611s
sys     0m0,238s
```
If I add the following here:
```
       for (auto const& name: m_variableScopes.back().variables)
       {
               m_value.erase(name);
               m_references.erase(name);
       }
```


vs.

```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m40,649s
user    0m40,308s
sys     0m0,241s
```
without adding it.

So no measurable difference - so I'll add it.

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: (for comparison: current develop is still at around 47 seconds:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m47,571s
user    0m46,993s
sys     0m0,357s
```
)

**Comment by ekpyron** [comment]
It would be nice, if someone tried to confirm that I'm right in that not clearing back-references when popping scopes is fine, since we check for variables being in scope on the user end (see https://github.com/ethereum/solidity/pull/10467#discussion_r533790328)

**Comment by ekpyron** [comment]
Nobody had another look at this ;-)? I just rebased it now, but it should still work unchanged and is still ready for review.

File: libsolutil/CommonData.h
Comment (leonardoalt) [inline libsolutil/CommonData.h]:
```suggestion
/// @returns a pointer to the entry of @a _map at @a _key, if there is one, and nullptr otherwise.
```

File: libyul/optimiser/CommonSubexpressionEliminator.cpp
Comment (leonardoalt) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp:93]: The semantics are not the same here, missing the assert.

File: libyul/optimiser/CommonSubexpressionEliminator.cpp
Comment (leonardoalt) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp:108]: Same here.

File: libyul/optimiser/Rematerialiser.cpp
Comment (leonardoalt) [inline libyul/optimiser/Rematerialiser.cpp]:
```suggestion
				bool allInScope = fold(m_references[name], true, [this](auto&& _acc, auto const& _ref) { return _acc && inScope(_ref); });
				solAssert(allInScope, "");
```

Comment (leonardoalt) [reply libyul/optimiser/Rematerialiser.cpp]: Loop can be removed then, but fine with either if you want to keep it.

File: libyul/optimiser/LoadResolver.cpp
Comment (leonardoalt) [inline libyul/optimiser/LoadResolver.cpp:74]: This is also somehow different. Before, if the location is storage but the key isn't found, it would try the `else if`. Now in that case it won't.

**Review by leonardoalt — COMMENTED** [review]
Some comments, still need to check the changes in `popScope`

File: libyul/optimiser/CommonSubexpressionEliminator.cpp
Comment (ekpyron) [reply libyul/optimiser/CommonSubexpressionEliminator.cpp:93]: Yeah, that's the point, see https://github.com/ethereum/solidity/pull/10467#discussion_r533790328

Comment (leonardoalt) [reply libyul/optimiser/CommonSubexpressionEliminator.cpp:93]: Ah ok

File: libyul/optimiser/LoadResolver.cpp
Comment (ekpyron) [reply libyul/optimiser/LoadResolver.cpp:74]: It would have entered the ``else if``, but since the location would be storage, ``_location == StoreLoadLocation::Memory`` would be false, i.e. the ``else if`` would not be entered in the end, i.e. we might as well not even enter it from the start.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (leonardoalt) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: It looks fine to me but I don't have a deep knowledge about the component. I think we should have a review from @chriseth or @hrkrshnn on this specific question.

File: libyul/optimiser/Rematerialiser.cpp
Comment (ekpyron) [reply libyul/optimiser/Rematerialiser.cpp]: ``fold`` is a bit harsh... I could add ``std::ranges::all_of`` to the cxx20 header, though...

Comment (leonardoalt) [reply libyul/optimiser/Rematerialiser.cpp]: cpp20 also has a better version of `accumulate`

File: libyul/optimiser/LoadResolver.cpp
Comment (leonardoalt) [reply libyul/optimiser/LoadResolver.cpp:74]: Yea ok

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (hrkrshnn) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: Not so sure about the effects. Maybe this will have an effect when the ast is not disambiguated. We saw earlier that DataFlowAnalyzer works well with non-disambiguated code. 

One example of this situation is:

```
  {
      {
          let a := 1
          let b := 2
          mstore(a, b)
      }
      {
          let a := 1
          let b := 3
          let bug := mload(a)
      }
  }
// ----
// step: loadResolver
//
// {
//     {
//         let a := 1
//         let b := 2
//         mstore(a, b)
//     }
//     {
//         let a := 1
//         let b := 3
//         let bug := b
//     }
// }
```

(The output is after removing disambiguator before loadresolver. [Here](https://gist.github.com/hrkrshnn/55f5303e4b6b2f1183818b8eb3eb4560) is the diff)

File: libyul/optimiser/Rematerialiser.cpp
Comment (ekpyron) [reply libyul/optimiser/Rematerialiser.cpp]: ``std::ranges::all_of`` is actually best here - it can even return early in some cases.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: That might very well be - but even if so, I wouldn't say it's a problem - we generally can assume that sources are disambiguated and we should be able to rely on them being disambiguated. The fact that we had a bug in that once and were lucky that we were not strongly hit by it, is not a good argument for dropping the assumption in general :-).

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: By the way: in general I don't think it works with disambiguated sources - the bug we had just only introduced name conflicts in very specific circumstances, which luckily never resulted in bad code, but in general I'd still expect non-disambiguated code to break anyways.

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp:301]: But yeah, it's a good catch, I think you're right, non-disambiguated sources would indeed have problems with this change... not sure that should stop us, though.

File: libsolutil/CommonData.h
Comment (hrkrshnn) [inline libsolutil/CommonData.h]: Do you think return an `optional<KeyType>` makes sense here? We could `move` the value at the callsite. May even be renamed to `optionalFind`.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (hrkrshnn) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: Maybe just expanding the `!` makes this look more pleasing? That is changing into `!a && !b`.

File: libyul/optimiser/LoadResolver.cpp
Comment (hrkrshnn) [inline libyul/optimiser/LoadResolver.cpp]: Does it make sense here to check if `value` is in current scope (`DataFlowAnalyzer::inScope`)? Maybe a bit safer.

File: libsolutil/cxx20.h
Comment (hrkrshnn) [inline libsolutil/cxx20.h]: Nice!

**Review by hrkrshnn — COMMENTED** [review]
Some tiny comments. Looks good.

File: libsolutil/CommonData.h
Comment (ekpyron) [reply libsolutil/CommonData.h]: What we'd need would be ``optional<ValueType&>`` - but unfortunately ``std::optional`` doesn't work with reference types (I think that's actually a standard defect report IIRC...)... the closest thing to it is just a pointer.
An actual ``optional<ValueType>`` (no reference) would always require a copy, so that's best avoided, so the plain pointer is probably the best we can do...

Comment (hrkrshnn) [reply libsolutil/CommonData.h]: The use cases so far will copy it anyway, right? So would it make a difference in the end?

Comment (ekpyron) [reply libsolutil/CommonData.h]: This function can be used all over the place throughout the compiler - I just didn't want to burden this PR with all that yet :-). There's 484 instances of ``.count(...)`` in the code-base - I'd estimate around half of those only use it as guard for accessing and will move to ``fetchValue``. A good part of those won't be stuff you'd want to copy.

File: libsolutil/cxx20.h
Comment (ekpyron) [reply libsolutil/cxx20.h]: Unfortunately, we can't use the more generic reference implementations in https://en.cppreference.com/w/cpp/algorithm/ranges/all_any_none_of, because they require concepts :-). But it should be enough like this for us.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Yeah, I thought about it, but since it was like this before, I just kept it - I'm not entirely sure which way it's easier to make sure the condition makes sense myself :-). But yeah, I can deMorgen-flip it.

File: libyul/optimiser/LoadResolver.cpp
Comment (ekpyron) [reply libyul/optimiser/LoadResolver.cpp]: Yep, probably better, added it.

**Comment by ekpyron** [comment]
By the way: the numbers have slightly changed since last time I measured - apparently ``develop`` has regressed a bit further performance-wise.
I now get on develop:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m55,125s
user    0m54,665s
sys     0m0,233s
```

vs. on this branch:

```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m47,796s
user    0m47,389s
sys     0m0,266s
```

which is still not much, but still roughly 15% speedup.

EDIT: also since @hrkrshnn asked: actually in the latest version changing the ``unordered_map``s back to regular ``map``s also does not make a huge difference, but it's a few percent slower, so since I don't see a reason for not using ``unordered_map``s I'd keep them.

**Review by hrkrshnn — DISMISSED** [review]
Looks good. A second approval would be nice before merging

**Review by leonardoalt — DISMISSED** [review]

**Comment by chriseth** [comment]
Haven't reviewed this in detail, but I'm not so sure about unordered_map. Can it be system-dependent? Also, about the performance comparisons: Please take into account that if you use isoltest, you are running this on many small files, which is not what we should optimize for.

**Comment by ekpyron** [comment]
Yeah, sure - it'd be nice to have a better set of sources to compare against... do we have that or want to come up with one?

About ``unordered_map``: do you mean whether the actual order in it is implementation-dependent or the performance? I guess for both the answer is yes, but nothing in the PR should depend on that order (I hope at least :-)).

That being said we might as well switch back to ``map`` - it doesn't make a huge difference in the last version of the PR.

**Comment by leonardoalt** [comment]
Well 15% is considerable difference

**Comment by axic** [comment]
> Yeah, sure - it'd be nice to have a better set of sources to compare against... do we have that or want to come up with one?

The deposit contract test case was merged, that is a good one to start with (https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/externalContracts/deposit_contract.sol)

**Comment by ekpyron** [comment]
> Well 15% is considerable difference

That it is, but we'll also get like 14% if we'd replace ``std::unordered_map`` with ``std::map`` and left anything else as is in this PR.

**Comment by ekpyron** [comment]
> > Yeah, sure - it'd be nice to have a better set of sources to compare against... do we have that or want to come up with one?
> 
> The deposit contract test case was merged, that is a good one to start with (https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/externalContracts/deposit_contract.sol)

```
$ time { for i in $(seq 100); do  buildOptimized/solc/solc --experimental-via-ir --optimize test/libsolidity/semanticTests/externalContracts/deposit_contract.sol &> /dev/null ; done ; }

real    0m30,724s
user    0m29,951s
sys     0m0,408s

$ time { for i in $(seq 100); do  build/solc/solc --experimental-via-ir --optimize test/libsolidity/semanticTests/externalContracts/deposit_contract.sol &> /dev/null ; done ; }

real    0m35,907s
user    0m35,369s
sys     0m0,425s
```

Rather consistent with the other measurements. But one contract doesn't really make for a good test set ;-).

**Comment by axic** [comment]
> Rather consistent with the other measurements. But one contract doesn't really make for a good test set ;-).

Well you can pick from #10390.

**Comment by ekpyron** [comment]
@hrkrshnn asked me to run https://gist.github.com/hrkrshnn/24818ae3ed9f1c6f04239f1765900501, so here:
```
$ time { for in in $(seq 10); do build/solc/solc --optimize --ir-optimized DexConstraintPoly.sol &> /dev/null ; done ; }

real    1m7,018s
user    1m6,774s
sys     0m0,123s

$ time { for in in $(seq 10); do buildOptimized/solc/solc --optimize --ir-optimized DexConstraintPoly.sol &> /dev/null ; done ; }

real    1m4,477s
user    1m4,216s
sys     0m0,140s
```

That's significantly less improvement, but @hrkrshnn also already pointed out that the file doesn't have a lot of control flow.

File: libsolutil/CommonData.h
Comment (chriseth) [inline libsolutil/CommonData.h]: Needs rebase.

Comment (ekpyron) [reply libsolutil/CommonData.h]: Ah, nice https://github.com/ethereum/solidity/pull/10739 was merged!

Comment (ekpyron) [reply libsolutil/CommonData.h]: Rebased.

**Comment by ekpyron** [comment]
Rebased.

File: libsolutil/cxx20.h
Comment (chriseth) [inline libsolutil/cxx20.h]:
```suggestion
template<class Key, class T, class Compare, class Alloc, class Pred>
```

File: libsolutil/cxx20.h
Comment (chriseth) [inline libsolutil/cxx20.h]:
```suggestion
typename std::map<Key, T, Compare,Alloc>::size_type erase_if(std::map<Key, T, Compare,Alloc>& c, Pred pred)
```

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (chriseth) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: I think it would be helpful to have a named utility function for this. It's just too easy to confuse things here.

Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: On the other hand, maybe it's better to go over `m_storage` just once and combine this with line 262?

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Sorry, but I'm not really sure what you mean :-). This line *is* calling a named utility function :-). You mean it's confusing that the "entry.second" is the value of the map, resp. the slot contents? To me it seems that extracting anything here would make this *harder* to read, not easier, but maybe I'm just not quite sure what you mean...

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (chriseth) [inline libyul/optimiser/DataFlowAnalyzer.cpp]:
```suggestion
	for (auto const& variableToClear: _variables)
```
(we have `name` and `names`, but they refer to completely different things)

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (chriseth) [inline libyul/optimiser/DataFlowAnalyzer.cpp]: `names` can now contain variables hat have gone out of scope, correct?

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: > On the other hand, maybe it's better to go over `m_storage` just once and combine this with line 262?

It's not the same kind of going over the it - the first one can take advantage of it being a hash and removes keys faster, yielding a smaller map to run over in the second one quickly. I actually tried and to my own surprise it was noticably slower to combine them - if I recall correctly at least. I could try again and remeasure, though...

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: On the other hand rather weird that the one entry more or less would actually make a difference, though... I'll recheck and see what happens...

Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Sure, there are more comparisons to be made! What I meant was that it is not obvious at first glance what this line does, and a helper function called `eraseValues` would make that clearer.

**Review by chriseth — COMMENTED** [review]
Looks good apart from style. I hope we can change the storage and memory tracking to only operate on SSA variables and then remove the complicated reference tracking at least for those.

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: I'd actually argue the opposite that this line makes it perfectly obvious what it does, because everything is right there, whereas such a helper function would mean that I need to check what it actually does first...

Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: ... with the difference that you already get a hint of what it is supposed to be doing. Using `.second` is always very confusing and I have to juggle with three things at a time (entry, name, second) to see what is going on.

Comment (chriseth) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Also: This is a lambda function with capturing, so the first thing I have to do is check what the variables in there are actually referencing in the current scope. If it's a free function, this problem is gone, or at least the scope is much smaller.

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: We could define something like
```
template<typename Callable>
decltype(auto) mapPair(auto&& callable) { return [&](auto&& pair) { return callable(pair.first, pair.second); }
```
And make it ``cxx20::erase_if(m_storage, mapPair([&name](YulString /* key */, YulString value) { return value == name; }));`` ;-)...

Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: That might solve the ``.second`` thing in more generality... and makes explicit that it doesn't really bind much...

File: libyul/optimiser/DataFlowAnalyzer.cpp
Comment (ekpyron) [reply libyul/optimiser/DataFlowAnalyzer.cpp]: Yep

File: libsolutil/CommonData.h
Comment (ekpyron) [inline libsolutil/CommonData.h:293]: So yeah, @chriseth: Weren't you actually complaining about ``[](auto&& [key, value]) { .. }`` not being possible recently :-)? This is a workaround for that :-)...

And yeah - look at the latest commit here, that's this in use - it actually helps in more cases than in that ``erase_if``...

But if you think this is too much and too complicated or still not maximally readable, just let me know, then I'll just drop that last commit and add an ``eraseValue`` helper instead, I really don't mind too much at this point :-).

**Review by chriseth — APPROVED** [review]
Still not convinced about `cxx20::erase_if(m_memory, mapTuple([&name](auto&& /* key */, auto&& value) { return value == name; }));`, but since we plan to remove `m_storage` and `m_memory` anyway, this is fine.

**Pull request merged**
